framework/src/modules/dpos_v2/auxiliary.ts

```diff
@@ -0,0 +1,64 @@
+import { BlockAfterExecuteContext } from '../../node/state_machine/types';
+import { MODULE_ID_DPOS, STORE_PREFIX_DELEGATE } from './constants';
+import { delegateStoreSchema } from './schemas';
+import { DelegateAccount, ModuleConfig, ValidatorsAPI } from './types';
+
+export const updateDelegateProductivity = async ({
+	validatorsAPI,
+	previousTimestamp,
+	context,
+	moduleConfig,
+}: {
+	validatorsAPI: ValidatorsAPI;
+	previousTimestamp: number;
+	context: BlockAfterExecuteContext;
+	moduleConfig: ModuleConfig;
+}) => {
+	const { header, getAPIContext, getStore } = context;
+
+	const newHeight = header.height;
+	const apiContext = getAPIContext();
+	const missedBlocks = await validatorsAPI.getGeneratorsBetweenTimestamps(
+		apiContext,
+		previousTimestamp,
+		header.timestamp,
+	);
+
+	const generatorAtPreviousTimestamp = await validatorsAPI.getGeneratorAtTimestamp(
+		apiContext,
+		previousTimestamp,
+	);
+	const generatorAtCurrentTimestamp = await validatorsAPI.getGeneratorAtTimestamp(
+		apiContext,
+		header.timestamp,
+	);
+
+	missedBlocks[generatorAtPreviousTimestamp.toString('hex')] -= 1;
+	missedBlocks[generatorAtCurrentTimestamp.toString('hex')] -= 1;
+
+	const delegateStore = getStore(MODULE_ID_DPOS, STORE_PREFIX_DELEGATE);
+	for (const addressString of Object.keys(missedBlocks)) {
+		const address = Buffer.from(addressString, 'hex');
+		const delegate = await delegateStore.getWithSchema<DelegateAccount>(
+			address,
+			delegateStoreSchema,
+		);
+		delegate.consecutiveMissedBlocks += missedBlocks[addressString];
```

Comment:
why we are incrementing missedBlocks here?

Reply:
Because `missedBlocks` array is iterable meaning it has values in it and some blocks are missed. If more than 1 block is missed since the last forge - that is recorded with previous timestamp - we add that value directly to consecutive missed blocks.